include/openenclave/host.h

```diff
@@ -64,6 +65,38 @@ typedef void (*oe_ocall_func_t)(
     size_t output_buffer_size,
     size_t* output_bytes_written);
 
+/**
+ * Types of configurations passed into oe_create_enclave
+ */
+typedef enum _oe_enclave_config_type
+{
+    OE_ENCLAVE_CONFIG_CONTEXT_SWITCHLESS = 0xdc73a628,
```

Comment:
I will say the same thing I say to any PR touching public header files...
I do not believe any PR should be merged that adds/changes a public API unless that PR comes with sample code and tests, or the API is an addition that is explicitly marked as experimental so breaking changes can be made at any time.  (This one comes with tests but no sample code, and no marking of experimental.)   I believe the simplest way to address for now is to add some way to mark this as experimental, e.g., by putting this inside an #ifdef OE_EXPERIMENTAL or some other annotation, with a // comment saying it's experimental and breaking changes may occur in the future once more experience is gained.

Reply:
We have created an issue to add sample code for switchless. https://github.com/openenclave/openenclave/issues/2130

Reply:
https://github.com/openenclave/openenclave/blob/master/docs/Contributing.md says:
"DON'T try to merge a giant feature in a single PR. Instead, break it into pieces with feature flags (pre-processor guards or --experimental flags), and submit multiple, smaller PRs to review each piece. Once the whole feature is in, agreed on, and tested, submit a PR to remove the guards."

I'm happy to +1 this PR once that's followed by adding a feature guard in the public header.  My OE_EXPERIMENTAL comment is about following the above requirement to have a guard since this is just a piece of a whole feature (sample being another piece).

Reply:
@dthaler Just to clarify, the public API changes in host.h can be split into two types:

1. The added structs related to switchless configuartions, which could be changed frequently.
2. The change to `oe_create_enclave`, where we changed the type of one parameter `configs` from `void*` to `oe_enclave_config_t`, and renamed a couple of other parameter names. This seems to be a stable change.

Do you want `OE_EXPERIMENTAL` to be applied on both or just the former?  If you want both, that means we need two versions of `oe_create_enclave` in this header.

Reply:
Only this file, and I would say yes to 1, and 2 is debatable... really oe_enclave_config_t needs a sample which is why I would say probably both.  But you don't need two full versions of oe_create_enclave per se.  For example,

oe_result_t oe_create_enclave( 
    const char* path,
    oe_enclave_type_t type,
    uint32_t flags,
#ifdef OE_CONTEXT_SWITCHLESS_EXPERIMENTAL_FEATURE
    const oe_enclave_config_t* configs,
    uint32_t config_count,
#else
    const void* config,
    uint32_t config_size,
#endif
    const oe_ocall_func_t* ocall_table,
    uint32_t ocall_table_size,
    oe_enclave_t** enclave);   

You then remove the #ifdefs in the PR that adds the sample (or later if you wish).

Reply:
If we keep two versions of oe_create_enclave, the tool oeedger8r has to generate two versions of calls into this function depending on whether `OE_CONTEXT_SWITCHLESS_EXPERIMENTAL_FEATURE` is on or off. How is that possible is beyond me since I am not familiar with the tool.

@anakrish @andschwa Could you help?

With the suggested changes, we need to modify a few places where oe_create_enclave does take an `oe_enclave_config_t*` pointer as parameter. Plus `host/sgx/create.c` in which we handle this parameter.

Reply:
@jxyang oeedger8r does not need to generate two versions; rather it just needs to generate a single version that compiles differently based on whether OE_CONTEXT_SWITCHLESS_EXPERIMENTAL_FEATURE is defined or not.

This PR already changes
https://github.com/openenclave/openenclave/blob/e7a8a23c136e88fb5960a6836afa43a6e19b1659/tools/oeedger8r/Emitter.ml#L1367-L1373

You just need to shove in the OE_CONTEXT_SWITCHLESS_EXPERIMENTAL_FEATURE as well:

```
    ; sprintf "oe_result_t oe_create_%s_enclave(" ec.enclave_name
    ; "    const char* path,"
    ; "    oe_enclave_type_t type,"
    ; "    uint32_t flags,"
    ; "#ifdef OE_CONTEXT_SWITCHLESS_EXPERIMENTAL_FEATURE"
    ; "    const oe_enclave_config_t* configs,"
    ; "    uint32_t config_count,"
    ; "#else"
    ; "    const void* config,"
    ; "    uint32_t config_size,"
    ; "#endif"
    ; "    oe_enclave_t** enclave);"
```

Similarly the signature at the function definition also needs to be changed.

Reply:
Cool. Thanks, Anand.

Reply:
great question.  Agree with Anand that's one way to do it (there may be others).   Since we're setting a precedent for other new features to follow, I'm in favor of variants that have the #ifdef in as few files as possible.  1 would be ideal.  Maybe 2 is the best we can do here.

Reply:
Thanks for the clarification. @dthaler.

Thanks for the suggestion, @anakrish.

Feature flag `OE_CONTEXT_SWITCHLESS_EXPERIMENTAL_FEATURE` is added.